### PR TITLE
render error page on 502 proxy status

### LINKED ISSUE #545: OAuth2 Proxy should return it's own error page when upstream connections fail
## Expected Behavior

If OAuth2 Proxy has an error of some description, it should return a rendered ErrorPage with details of why the error occurred

## Current Behavior

If the OAuth2 Proxy cannot connect to the upstream to serve the request, an empty response with a 502 is returned. This implies that the connection to OAuth2 Proxy failed instead of OAuth2 Proxy's connection to the upstream 

```
2020/05/12 11:13:37 http: proxy error: dial tcp: lookup httpbin on 127.0.0.11:53: no such host
172.20.0.1 - admin@example.com [2020/05/12 11:13:32] localhost:4180 GET httpbin "/anything" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.129 Safari/537.36" 502 0 5.028
```

## Possible Solution

Set the `ErrorHandler` on the `ReverseProxy` to render our `ErrorPage` when there are connection errors to upstream servers 

Ref: https://golang.org/pkg/net/http/httputil/#ReverseProxy

## Steps to Reproduce (for bugs)

This requires the test environment I've set up in #529 

1.  `make local-env-up` <- Brings up docker environment
2.  Perform login process by accessing `http://localhost:4180` (instructions in files in #529)
3.  Reach HTTPBin succeessfully
4.  `docker stop httpbin`
5. Access any URL, naked 502 is returned

## Context

In doing this, we should improve the UX of the project and help users narrow down errors that they are seeing

## Your Environment

- Version used: 5.1.1 (Present in master)

### PR DESCRIPTION
**Pull request by amnay-mo**
## Description

Render an error page with detailed error on proxy 502 status
Fixes #545 

## Motivation and Context

A blank 502 page is displayed whenever the reverse proxy can't get a valid response from an upstream

## How Has This Been Tested?

I Tested using the local environment provided in the contrib files (the httpbin + dex setup)

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: oauthproxy.go [hunk 1]
```diff
@@ -154,9 +154,28 @@ func NewReverseProxy(target *url.URL, opts *options.Options) (proxy *httputil.Re
...
+	proxy.ErrorHandler = func(w http.ResponseWriter, r *http.Request, proxyErr error) {
+		logger.Printf("proxying error :%v", proxyErr)
+		w.WriteHeader(http.StatusBadGateway)
+		data := struct {
+			Title       string
+			Message     string
+			ProxyPrefix string
+		}{
+			Title:       "Bad Gateway",
+			Message:     fmt.Sprintf("could not get response from upstream '%s': %v", target.String(), proxyErr),
```
Comment (JoelSpeed) [inline oauthproxy.go:172]: I'm unsure if we want to print the target, that could reveal internal details about the systems, what do you think?

Comment (amnay-mo) [reply oauthproxy.go:172]: Good point!
But we could still leak internal network details by printing out the proxy error (such as the ip address and port of the upstream in the case of a TCP error)
What do you think?

Comment (JoelSpeed) [reply oauthproxy.go:172]: Don't suppose you happen to have an example of the printed logs from your manual testing? Seeing the real log would help make the decision I think.

If not I'll spin up an env later and have a try

Comment (amnay-mo) [reply oauthproxy.go:172]: Here, the upstream is unreachable:

image

Comment (JoelSpeed) [reply oauthproxy.go:172]: I'd be tempted to not display the errors to the end user, instead maybe just display

`Error proxying to upstream server`

The user should be expected to report the problems to the administrator right? And they would have access to the logs in which they should find the logging of the errors right?

Comment (amnay-mo) [reply oauthproxy.go:172]: I completely agree with you
Thanks for the feedback! :+1:

**Pull request merged**